File: 2.add-two-numbers/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var sum int
	var head ListNode
	l3 := &head
	for l1 != nil || l2 != nil {
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		l3.Next = new(ListNode)
		l3 = l3.Next
		l3.Val = sum % 10
		sum /= 10
	}
	if sum != 0 {
		l3.Next = new(ListNode)
		l3 = l3.Next
		l3.Val = sum % 10
	}
	return head.Next
}

File: 2.add-two-numbers/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const p = new ListNode();
    let q = p;
    let c = 0, sum = 0;
    while (l1 || l2 || c) {
        sum = (l1?.val ?? 0) + (l2?.val ?? 0) + c;
        q.next = new ListNode(sum % 10);
        q = q.next;
        c = (sum - q.val) / 10;
        l1 = l1?.next;
        l2 = l2?.next;
    }
    return p.next;
};
